Remove AlignDetectors algorithm

sf1919 opened the pull request:
### Description of work

Removal of AlignDetectors Algorithm as this has been replaced with ConvertUnits.

#### Summary of work
<!-- Please provide a short, high level description of the work that was done.
-->

<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
#### Purpose of work
This can be removed if a github issue is referenced below
-->

Part of #31258


#### Further detail of work
Deprecated in 2021 - see #30163 

### To test:

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

sf1919 committed:
Remove AlignDetectors algorithm

- Remove head and main files
- Remove references from CMake

sf1919 committed:
Remove aligndetectorstest

- Remove test and reference from CMake

sf1919 committed:
Remove user documentation

- Removed the User documentation of the algorithm
- Removed links from previous release notes
- Removed references to the AlignDetectors Algorithm as an example in documentation

sf1919 committed:
Remove references for AlignDetectors in algorithm documentation

- Where possible remove or convert reference to AlignDetectors to ConvertUnits

sf1919 committed:
Update algorithm documentation with maths from AlignDetectors

- For some algorithm documentation there are references made to the GSAS equations in AlignDetectors. This commit moves that information into the relevant documentation to enable the references to AlignDetectorsto be removed.

sf1919 commented:
@RichardWaiteSTFC can you examine the commits [c3b9e0d](https://github.com/mantidproject/mantid/pull/37799/commits/c3b9e0d3efaf037fb9ddcd5b3ed0846f6d9f3b05) and [b25ac52](https://github.com/mantidproject/mantid/pull/37799/commits/b25ac52da2cc2fc4297db0b0174331a30e4a1a5d) please? I'm not sure if I've made quite the right adjustments in the documentation on these.

RichardWaiteSTFC commented on docs/source/algorithms/EnggFitPeaks-v1.rst:
```diff
@@ -33,11 +33,38 @@ then the algorithm fails.
+:math:`TZERO` is related to the difference between the measured and
+actual time-of-flight base on emission time from the moderator, :math:`DIFA` is an empirical term (ideally zero), and :math:`DIFC` is
+
+.. math:: DIFC = \frac{2m_N}{h} L_{tot} sin \theta
```
I think this info should actually live in unit factory docs
https://docs.mantidproject.org/nightly/concepts/UnitFactory.html#unit-factory
or in the `ConvertUnits` docs.
This algorithm is no longer used (it pre-dates my work on engineering GUI) - also the name suggest something engineering specific whereas this relates to TOF powder diffraction.

sf1919 replied:
Does ConvertUnits actually use `DIFC`? I can't see an obvious place to put it in the documentation.

RichardWaiteSTFC commented on docs/source/algorithms/PDCalibration-v1.rst:
```diff
@@ -43,7 +43,7 @@ When not specified using the ``MaskWorkspace`` parameter, the default name for t
 The resulting calibration table can be saved with
 :ref:`algm-SaveDiffCal`, loaded with :ref:`algm-LoadDiffCal` and
-applied to a workspace with :ref:`algm-AlignDetectors`. There are also
+applied to a workspace with :ref:`algm-ConvertUnits`. There are also
```
I think we should also mention `ApplyDiffCal` - which actually uses the calibration table and updates the diffractometer constants in the `UnitParams` map of the workspace

sf1919 committed:
Move GSAS equation information

- The GSAS equation information would be best situated in the Unit Factory concept documentation. This commit makes that change and updates references to this concept documentation instead.

sf1919 committed:
Add mention of ApplyDiffCal algorithm in PDCalibration documentation

sf1919 committed:
Update calibration concept documentation

- Remove references to AlignDetectors from CalFile and PowderDiffractionCalibration.
- Replace references where necessary with suitable alternatives

sf1919 committed:
Update DiffractionFocussing2

- DiffractionFocussing2 made reference to using AlignDetectors. This has been changed to ConvertUnits
- The associated test used AlignDetectors. This has been changed to ConvertUnits and variables renamed accordingly
- Updated DiffractionFocussing2 documentation

sf1919 committed:
Add ApplyDIffCal to DiffractionFocussing2

- ConvertUnits is not a direct swap with AlignDetectors. Instead a combination of ApplyDIffCal and ConvertUnits is needed. This makes that fix,

sf1919 commented:
Currently blocked by #37859

sf1919 committed:
Remove see also references

Remove links from other algorithms for AlignDetectors under the 'see also' section

sf1919 committed:
Update the gem_routines/gem_calibration_algs script

- Remove AlignDetectors and replace with ApplyDiffCal and COnvertUnits algorithms

sf1919 committed:
Remove AlignDetectors from WishAnalysis System test

- Replaced with ApplyDiffCal and ConvertUnits algorithms

sf1919 committed:
Use correct parameters for ApplyDiffCal

- ApplyDiffCal does not take an InputWorkspace. Instead it takes an InstrumentWorkspace.

sf1919 committed:
More documentation cleaning

- Removed link to alignDetectors from See Also in SaveGDA
- Removed reference from explanation of OffsetsWorkspace

sf1919 commented:
Some workflow diagrams refer to `AlignDetectors` algorithm. However these appear to be generally out of date and so separate issues have been raised to deal with these (see #37905 and #37907)

sf1919 committed:
Fix problems with WishAnalysis test

sf1919 committed:
Fix failing system tests

sf1919 committed:
Add release note

sf1919 commented:
I'm moving this to 6.12 as there are some issues that I think need to be fixed separately before this one can continue. I'll write up the relevant issues.

sf1919 committed:
Upate imports in WishAnalyis.py test

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

sf1919 committed:
Move release note

RichardWaiteSTFC commented on Testing/SystemTests/tests/framework/WishAnalysis.py:
```diff
@@ -64,9 +64,8 @@ def runTest(self):
-            InputWorkspace="w16748-1", OutputWorkspace="w16748-1", CalibrationFile="wish_grouping_noends2_no_offsets_nov2009.cal"
-        )
+        ApplyDiffCal(InstrumentWorkspace="w16748-1", CalibrationFile="wish_grouping_noends2_no_offsets_nov2009.cal")
+        ConvertUnits(InputWorkspace="w16748-1", OutputWorkspace="w16748-1", Target="dSpacing")
```
Think this test reference workspace is actually converted back to TOF so need to run `ApplyDiffCal` with `ClearCalibration=True` after L68 (or anywhere before the call to `ConvertUnits` back to TOF

RichardWaiteSTFC committed:
Fix WISH system test by clearing calibration

Reference file was converted back to TOF so for backwards compatibility need to clear calibration (so uncalibrated FIC used for backward conversion)

RichardWaiteSTFC committed:
Fix gem calibration test by clearing calibration

Even though the reference is the focussed d-spacing spectrum (so no conversion back to TOF required), the test was failing because the parameter map in the two workspaces was different.

RichardWaiteSTFC committed:
Fix doc test still using AlignDetectors

RichardWaiteSTFC committed:
Remove references to AlignDetectors in docs and comments etc.

RichardWaiteSTFC commented on docs/source/concepts/calibration/PowderDiffractionCalibration.rst:
```diff
@@ -662,13 +659,13 @@ This can be done using the following code
    Rebin(InputWorkspace='ws', OutputWorkspace='ws', Params=(5000,-.002,70000))
-   LoadDiffCal(Filename='VULCAN_Calibration_CC_4runs_hybrid.h5', InputWorkspace='ws', WorkspaceName='VULCAN')
-   AlignDetectors(InputWorkspace='ws', OutputWorkspace='ws', CalibrationWorkspace='VULCAN_cal')
+   ApplyDiffCal(InputWorkspace='ws', CalibrationFile='VULCAN_Calibration_CC_4runs_hybrid.h5')
```
@wdzhou - I don't have the files to test this change and it's not covered by a doc test, but this should work as it looks like  `ApplyDiffCal` calls `LoadDiffCal` - do you agree? Don't know if you're still working on VULCAN calibration and may have these files anywhere?

cailafinn replied:
Did we determine an answer to this question?

peterfpeterson replied:
Looking at the lines of change, this looks identical to what we did in other places.

Wenduo left the project a couple of years ago so don't expect a response from him.

RichardWaiteSTFC commented on scripts/Diffraction/isis_powder/gem_routines/gem_calibration_algs.py:
```diff
@@ -138,7 +138,9 @@ def _calibration_processing(
     )
     rebinned_tof = mantid.ConvertUnits(InputWorkspace=rebinned, Target="TOF")
-    aligned = mantid.AlignDetectors(InputWorkspace=rebinned_tof, CalibrationFile=offset_file)
+    mantid.ApplyDiffCal(InstrumentWorkspace="rebinned_tof", CalibrationFile=offset_file)
```
Here it would actually be better to use the offsets workspace (`OffsetsWorkspace=offset_ws_name` here). The act of writing to a fixed width column can truncate a number, which in this case can lead to quite large deviations (relative to typical absolute tolerances used in system tests) because the y-values are large and the peak is quite sharp.
<img width="900" alt="image" src="https://github.com/user-attachments/assets/ae78bb5b-8997-4e4e-b309-29a2830d588f" />
Even though the d-spacing of the workspaces calibrated using the file and the offsets workspace differ only by ~1e-8

RichardWaiteSTFC replied:
I have kept the use of the offsets file so as to 'fix' the test and preserve the behaviour - in practice this difference in offset used will never be noticed in the refinement, it is only noticeable in system tests.

andy-bridger commented on docs/source/concepts/calibration/PowderDiffractionCalibration.rst:
```diff
@@ -597,7 +594,7 @@ The easiest way to do this is to apply the calibration to your
 
 1. Load the calibration data using :ref:`Load <algm-Load>`
-2. Run :ref:`AlignDetectors <algm-AlignDetectors>`, this will convert the data to d-spacing and apply the calibration.  You can provide the calibration using the ``CalibrationFile``, the ``CalibrationWorkspace``, or ``OffsetsWorkspace``.
+2. Run :ref:`ConvertUnits <algm-ConvertUnits>`, this will convert the data to d-spacing and apply the calibration.  You can provide the calibration using the ``CalibrationFile``, the ``CalibrationWorkspace``, or ``OffsetsWorkspace``.
```
Is this true? I thought ConvertUnits just converted it d-spacing and the calibration would have to be applied with `ApplyDiffCal`?

RichardWaiteSTFC replied:
I agree I think there's a missing step to call `ApplyDiffCal` which updates the parameter map on the workspace (i.e. the information used to convert between TOF -> d-spacing) then `ConvertUnits` will use this calibration. Previously `AlignDetectors` would apply the calibration to the conversion from TOF -> d-sapcing but there would be no update of the parameters on the workspace, such that the inverse conversion (from d-spacing -> TOF) would use the un-calibrated DIFC parameter.

andy-bridger replied:
I think it would also be nice to have that clarified in the docs (`docs/source/algorithms/ConvertUnits-v1.rst`) because I don't think it currently mentions anything about calibration

andy-bridger reviewed (changes requested):
Not sure how the review functionality works so whether this needs to be separate, but if we could maybe get these changes (https://github.com/mantidproject/mantid/pull/37799#discussion_r1941018708) to `docs/source/concepts/calibration/PowderDiffractionCalibration.rst` and `docs/source/algorithms/ConvertUnits-v1.rst`

RichardWaiteSTFC committed:
Add calibration section to ConvertUnits

Also clarify role of ConvertUnits and ApplyDiffCal in other places in docs

RichardWaiteSTFC commented:
Thanks @andy-bridger for the review, I've added the section you requested in `ConvertUnits` and clarified the role of `ApplyDiffCal` in `PowderDiffractionCalibration.rst` and a few other places that I could find.

Just a note when reviewing, anywhere in the code where `AlignDetectors` has been replaced there needs to be two calls to `ApplyDiffCal` - the first with the calibration file/workspace, the second call after `ConvertUnits` with `ClearCalibration=True` - this means the calibration won't be used to convert back from d-spacing to TOF (which was the behaviour of `AlignDetectors`).
I think I fixed this when I fixed the systems tests, but I'd appreciate a second pair of eyes on it!

andy-bridger commented on Framework/Algorithms/test/DiffractionFocussing2Test.h:
No second call of `ApplyDiffCal` here, but I'm guessing that is fine because it is just a test and there is no further processing

RichardWaiteSTFC replied:
Yep I think that's OK because it's checking the focussing produces a workspace with a single histogram - technically the DIFC of the focussed spectrum will be different but that's not part of what is being tested there

andy-bridger commented on docs/source/algorithms/DiffractionFocussing-v2.rst:
again I assume it is intentional to the example on L83 that `ApplyDiffCal` is not called with `ClearCalibration`

RichardWaiteSTFC replied:
Yep I think that's OK as well because the data are not converted back to TOF in that usage example

andy-bridger reviewed (dismissed):
All looks good now

cailafinn reviewed (changes requested):
Changes looking good, just a couple of quick Q's about the documentation around them:

sf1919 commented:
It doesn't look like we've heard back about VULCAN so I've asked @peterfpeterson for help

RichardWaiteSTFC commented on Framework/Algorithms/inc/MantidAlgorithms/DiffractionFocussing2.h:
```diff
@@ -74,9 +74,7 @@ class MANTID_ALGORITHMS_DLL DiffractionFocussing2 final : public API::Algorithm
-  const std::vector<std::string> seeAlso() const override {
-    return {"AlignDetectors", "AlignAndFocusPowder", "LoadCalFile"};
-  }
+  const std::vector<std::string> seeAlso() const override { return {"AlignAndFocusPowder", "LoadCalFile"}; }
```
I don't mind but IMO I don't think so - I guess it depends on what `seeALso` is meant to contain. I thought it was for related, similar algorithms (e.g. `FindUBUingLatticeParameters` would point to `FindUBUsingIndexedPeaks` etc.). `DiffractionFocussing` only works with some x-units but the act of summing up the data in detectors is not actually related to the unit conversion (I would have questioned `AlignDetectors` being in the `seeAlso` in the first place - it's probably because it's in the previous step in the powder diffraction reduction, but of course `ConvertUnits` is used in many more places than that)

RichardWaiteSTFC commented on docs/source/algorithms/ConvertUnits-v1.rst:
```diff
@@ -35,6 +35,22 @@ If ConvertFromPointData is true, an input workspace
+(from :ref:`GetDetectorOffsets <algm-GetDetectorOffsets>`) or diffractometer constants
+(from :ref:`PDCalibration <algm-PDCalibration>`) for the conversion between TOF and d-spacing.
+For more details on diffractometer constants see :ref:`Unit Factory`, for the relation between offsets and diffractometer
+constants see ref:`ConvertDiffCal <algm-ConvertDiffCal>`.
```
OK thanks, didn't know that!

RichardWaiteSTFC commented:
> It doesn't look like we've heard back about VULCAN so I've asked @peterfpeterson for help

`ApplyDiffCal` calls `LoadDiffCal` if a calibration file is provided so I'm pretty confident it's a like-for-like swap!
https://github.com/mantidproject/mantid/blob/d1e96c3f6567e1b4b4dd64c89c1e0ff480a4ef94/Framework/DataHandling/src/ApplyDiffCal.cpp#L150-L154
https://github.com/mantidproject/mantid/blob/d1e96c3f6567e1b4b4dd64c89c1e0ff480a4ef94/Framework/DataHandling/src/ApplyDiffCal.cpp#L122-L123

The only thing that might be an issue is that the calibration if not cleared subsequently, however it looks like there are no future conversions back to TOF in the file so this doesn't matter. On second thought we might as well do it!

peterfpeterson reviewed (commented):
I'm happy if this is merged. I didn't approve so as to not step on toes.

sf1919 committed:
Update docs/source/algorithms/ConvertUnits-v1.rst

Co-authored-by: Caila Finn <47181718+cailafinn@users.noreply.github.com>
